x-pack/plugins/task_manager/task_store.ts

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/*
+ * This module contains helpers for managing the task manager storage layer.
+ */
+
+import { ConcreteTaskInstance, ElasticJs, TaskInstance, TaskStatus } from './task';
+
+const DOC_TYPE = '_doc';
+
+export interface StoreOpts {
+  callCluster: ElasticJs;
+  index: string;
+  maxAttempts: number;
+  supportedTypes: string[];
+}
+
+export interface FetchOpts {
+  searchAfter?: any[];
+  sort?: object[];
+  query?: object;
+}
+
+export interface FetchResult {
+  searchAfter: any[];
+  docs: ConcreteTaskInstance[];
+}
+
+export interface RemoveResult {
+  index: string;
+  id: string;
+  version: string;
+  result: string;
+}
+
+// Internal, the raw document, as stored in the Kibana index.
+export interface RawTaskDoc {
+  _id: string;
+  _index: string;
+  _type: string;
+  _version: number;
+  _source: {
+    type: string;
+    task: {
+      taskType: string;
+      runAt: Date;
+      interval?: string;
+      attempts: number;
+      status: TaskStatus;
+      params: string;
+      state: string;
+      user?: string;
+      scope?: string | string[];
+    };
+  };
+}
+
+/**
+ * Wraps an elasticsearch connection and provides a task manager-specific
+ * interface into the index.
+ */
+export class TaskStore {
+  public readonly maxAttempts: number;
+  private callCluster: ElasticJs;
+  private index: string;
+  private supportedTypes: string[];
+  private wasInitialized = false;
+
+  /**
+   * Constructs a new TaskStore.
+   * @param {StoreOpts} opts
+   * @prop {CallCluster} callCluster - The elastic search connection
+   * @prop {string} index - The name of the task manager index
+   * @prop {number} maxAttempts - The maximum number of attempts before a task will be abandoned
+   * @prop {string[]} supportedTypes - The task types supported by this store
+   */
+  constructor(opts: StoreOpts) {
+    this.callCluster = opts.callCluster;
+    this.index = opts.index;
+    this.maxAttempts = opts.maxAttempts;
+    this.supportedTypes = opts.supportedTypes;
+  }
+
+  public addSupportedTypes(types: string[]) {
+    if (!this.wasInitialized) {
+      this.supportedTypes = this.supportedTypes.concat(types);
+    } else {
+      throw Error('Cannot add task types after initializattion');
+    }
+  }
+
+  /**
+   * Initializes the store, ensuring the task manager index is created and up to date.
+   */
+  public async init() {
+    if (this.wasInitialized) {
+      return;
```

Comment:
Attempting to double-init on something like this seems like broken behavior that should have an error thrown:

```ts
throw new Error('TaskStore has already been initialized!');
```

Reply:
f05e59f55a